### DISCUSSION TITLE
databend-query can't distinguish error meta service endpoints？

### DISCUSSION THREAD

**Discussion by lddlww** [databendlabs/databend#9768] [Q&A] [answered]
**databend version:**
```
v0.9.16-nightly-3fcebe7(rust-1.68.0-nightly-2023-01-27T16:29:17.399818523Z)
```
**my databend-query config toml with two error meta service endpoints：**
```
[meta]
endpoints = ["10.5.16.18:9191","10.5.16.18:87:9191","10.5.16.18:49:9191"]
username = "root"
password = "root"
client_timeout_in_second = 60
auto_sync_interval = 60

[log.file]
level = "INFO"
format = "text"
dir = "/data/databend/9000/log"
```
start databend-query normal using above config without any warn or error

**databend-query log like follow:**
```
2023-01-29T10:27:17.955203Z  INFO common_meta_store: src/meta/store/src/lib.rs:124: use remote meta conf=RpcClientConf { endpoints: ["10.5.16.18:9191", "10.5.16.18:87:9191", "10.5.16.18:49:9191"], username: "root", password: "root", tls_conf: None, timeout: Some(60s), auto_sync_interval: None }
2023-01-29T10:27:17.955615Z  INFO worker_loop: common_meta_client::grpc_client: src/meta/client/src/grpc_client.rs:341: MetaGrpcClient::worker spawned
```

**correct meta service endpoints**
```
["10.5.16.18:9191","10.5.16.87:9191","10.5.16.49:9191"]
```

**problem：**
when i leaved  ```10.5.16.18:9191``` meta service from correct meta service cluster , the databend-query can't be accessed and mysql client throw ```ERROR 1815 (HY000): endpoints is empty```

**Comment by bohutang** [comment]
Ping @drmingdrmer

**Comment by drmingdrmer** [comment, marked as answer]
Did you mean `leave` a command `./target/debug/databend-meta --leave-via`? 

Let me have a test on my laptop first.

**Reply by lddlww** [reply]
sorry，what i said my be misleading， what i mean is when i use error config like  ```endpoints = ["10.5.16.18:9191","10.5.16.18:87:9191","10.5.16.18:49:9191"]``` in databend-query config , **it can't throw any warn or error**

above endpoints only "10.5.16.18:9191"  is valid , so i leaved "10.5.16.18:9191"  using ```./target/debug/databend-meta --leave-via``` the databend-query can not work anymore

"10.5.16.18:87:9191","10.5.16.18:49:9191" is hands-on write error，so i want databend-query could remind me the error

**Reply by drmingdrmer** [reply]
And what's the address list of the correct databend-meta cluster?

**Reply by drmingdrmer** [reply]
Ok I see your problem!

**Reply by drmingdrmer** [reply]
If the databend-meta cluster has only one member "10.5.16.18:9191", then removing it will cause the query to be unable to run.

Such behavior should be forbidden. I'll fix this, then.

And I think `ERROR 1815 (HY000): endpoints is empty` is a correct error in such a case: there are no reachable endpoints

**Reply by lddlww** [reply]
> Such behavior should be forbidden. I'll fix this, then
thank you very much

> And I think `ERROR 1815 (HY000): endpoints is empty` is a correct error in such a case: there are no reachable endpoints
yes，you are right

**Reply by drmingdrmer** [reply]
I've merged a fix that forbids removing the last node in a cluster:
-  https://github.com/datafuselabs/databend/pull/9781

**Reply by lddlww** [reply]
that's great work
